etcdserver/server.go

```diff
@@ -1653,10 +1655,47 @@ func (s *EtcdServer) mayPromoteMember(id types.ID) error {
 	if !s.Cfg.StrictReconfigCheck {
 		return nil
 	}
-	// TODO add more checks whether the member can be promoted.
-	// like learner progress check or if cluster is ready to promote a learner
-	// this is an example to get progress
-	fmt.Printf("raftStatus, %#v\n", raftStatus())
+	if !s.cluster.IsReadyToPromoteMember(uint64(id)) {
+		if lg := s.getLogger(); lg != nil {
+			lg.Warn(
+				"rejecting member promote request; not enough healthy members",
+				zap.String("local-member-id", s.ID().String()),
+				zap.String("requested-member-remove-id", id.String()),
+				zap.Error(ErrNotEnoughStartedMembers),
+			)
+		} else {
+			plog.Warningf("not enough started members, rejecting promote member %s", id)
+		}
+		return ErrNotEnoughStartedMembers
+	}
+
+	// check learner progress is almost catching up the leader.
+	raftStatus := raftStatus()
+
+	// leader's raftStatus.Progress is not nil
+	if raftStatus.Progress == nil {
+		return ErrNotLeader
+	}
+
+	var learnerMatch uint64
+	isFound := false
+	leaderID := raftStatus.ID
+	for memberID, progress := range raftStatus.Progress {
+		if uint64(id) == memberID {
```

Comment:
Looks like during copying the progress, we lost information about whether the progress is for a learner or voting member. So trying to promote a voting member could return "learner not ready" error. I think we can live with this corner case for now, unless there is an easy fix (do we want to check if member is learner here using cluster information)?

Reply:
>So trying to promote a voting member could return "learner not ready" error. I think we can live with this corner case for now

yeah, now this kind of error is checked during the validation phase. 

>do we want to check if member is learner here using cluster information)

At the first implementaion I just use this approach while you said it maybe race condition.... 
I and xiang90 agreed it's ok now because the race is rare...